src/LocateUri.cxx

```diff
@@ -62,7 +63,7 @@ LocateAbsoluteUri(const char *uri
 		  )
 {
 	if (!uri_supported_scheme(uri))
-		throw std::runtime_error("Unsupported URI scheme");
+		throw ProtocolError(ACK_ERROR_NO_EXIST, "Unsupported URI scheme");
```

Comment:
The problem here is not that this URI doesn't exist (it does exist, after all!), but is an invalid URI. This sounds like `ACK_ERROR_ARG` is the correct code.

Reply:
Oh right, I just had this in my fork when testing because it made my life easier, as it's impossible to get ncmpcpp to execute a playlist load without throwing this error code back at it. I'll do up a separate PR for that change. That "try add first then load" behaviour is also part of the reason why the input plugin and the playlist plugin use two different URI schemes - not sure if that should be the case, I'll add that as a discussion point.

Reply:
So this is implemented as a playlist plugin only because you figured this allows you to attach metadata? No, that's not acceptable.
If you want to add one file, `mpc add` must be used, not `mpc load`. Playlists are for playlists.
This should be an input plugin, not a playlist plugin at all. If you want metadata, then yes you can implement a `RemoteTagScanner`.
Note that your code should be non-blocking to make it cancellable. Without this, MPD may block indefinitely when the server is not reachable. (This is a serious problem, and in my ongoing fight against existing blocking code, I don't want to merge any more new blocking code.)
For examples on how to do non-blocking input plugins with HTTP and JSON, look at the Tidal/Qobuz plugins.

Reply:
Well, it's a playlist plugin because it adds playlists too. It just happens that adding a single file playlist `load` pulls in the metadata immediately, which is a bit better UX than `add`ing the URL and not having it populate until the file is actually played. Will implementing a `RemoteTagScanner` resolve that?

Reply:
"because it adds playlists too" that's a different feature which requires implementing a different plugin. I think the most important feature is playing one single file by its id. You can add a playlist feature later, but that's not the primary goal.
With "populate", you mean provide metadata? Yes, `RemoteTagScanner` can do that, but it's an input plugin feature.

Reply:
I will never add kludges to MPD just because some client decided to make funny assumptions on something that's not explicitly documented in the protocol spec (and yes, the protocol spec isn't very specific on some things, and needs to improve).
I'm not sure if I should call this ncmpcpp behavior "bug", but I wouldn't have done that. I maintain ncmpc (of which ncmpcpp is a clone with a very unfortunate name), and ncmpc doesn't do it. Andd I maintain mpc, which doesn't do that either.

Reply:
Mm, not so much a bug as a missing feature - it's unable to load playlists, but just happens to be able to if the server returns the right error when adding a file!

On an unrelated note perhaps I should look into whether I can use ncmpc for personal use.